### PR TITLE
linuxPackages.rtl88xxau-aircrack: fix build for linux 5.15

### PR DESCRIPTION
**Pull request by MatthewCroughan**
Co-authored-by: nixinator <33lockdown33@protonmail.com>

ZHF: https://github.com/NixOS/nixpkgs/issues/144627

@jonringer, @tomberek , @nrdxp

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nixinator** [comment]
it compiles now after applying a upstream patch, don't have the hardware so can't test it running. Maybe some can do that?

**Review by SuperSandro2000 — COMMENTED** [review]
Why are we not bumping the package?

**Comment by MatthewCroughan** [comment]
> Why are we not bumping the package?

I saw the pattern of using the patch and decided to continue with it, just to get it compiling. Should we bump the version instead?

**Comment by MatthewCroughan** [comment]
@SuperSandro2000 

The latest tagged release does compiles successfully with a 5.15 kernel, though only if we remove the 5.12 kernel patch, so I have done this.

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/145697

2 packages marked as broken and skipped:
linuxKernel.packages.hardkernel_4_14.rtl88xxau-aircrack linuxPackages_hardkernel_latest.rtl88xxau-aircrack

1 package failed to build:
linuxPackages_hardened.rtl88xxau-aircrack

18 packages built:
linuxKernel.packages.linux_4_14.rtl88xxau-aircrack linuxPackages_4_14_hardened.rtl88xxau-aircrack linuxKernel.packages.linux_4_19.rtl88xxau-aircrack linuxPackages_4_19_hardened.rtl88xxau-aircrack linuxKernel.packages.linux_4_4.rtl88xxau-aircrack linuxKernel.packages.linux_4_9.rtl88xxau-aircrack linuxPackages.rtl88xxau-aircrack linuxKernel.packages.linux_5_14.rtl88xxau-aircrack linuxPackages_5_14_hardened.rtl88xxau-aircrack linuxPackages_latest.rtl88xxau-aircrack linuxKernel.packages.linux_5_4.rtl88xxau-aircrack linuxPackages_5_4_hardened.rtl88xxau-aircrack linuxPackages_latest-libre.rtl88xxau-aircrack linuxPackages-libre.rtl88xxau-aircrack linuxPackages_lqx.rtl88xxau-aircrack linuxPackages_testing_bcachefs.rtl88xxau-aircrack linuxPackages_xanmod.rtl88xxau-aircrack linuxPackages_zen.rtl88xxau-aircrack
```

cc @SuperSandro2000

File: pkgs/os-specific/linux/rtl88xxau-aircrack/default.nix [hunk 1]
```diff
@@ -20,14 +20,6 @@ stdenv.mkDerivation rec {
 
   NIX_CFLAGS_COMPILE="-Wno-error=incompatible-pointer-types";
 
-  # Fix build for 5.12 kernels
-  patches = [
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/rtl88xxau-aircrack/default.nix:25]: Please remove the unused fetchpatch from inputs.

**Comment by nixinator** [comment]
there are quite a few of these in out of tree wireless drivers. (due 5.15 kernel maintainers removing ipx for ever :-( )

https://hydra.nixos.org/build/157498591/nixlog/1

so they all these drivers need bumping/patching too :-).

how do we handle older kernels like 5.10.x , or is a 'nixos release' always tied to a 'support kernel'  and out of tree drivers needs to follow the 'nixos release kernel version'????

File: pkgs/os-specific/linux/rtl88xxau-aircrack/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/os-specific/linux/rtl88xxau-aircrack/default.nix:25]: https://github.com/NixOS/nixpkgs/pull/145808

**Comment by nixinator** [comment]
linuxPackages_5_15.rtl8812au

is source from https://github.com/gordboy/rtl8812au-5.9.3.2

however that doesn't compile of 5.15 , however 

https://github.com/morrownr/8812au-20210629

apparently does, however , do we swap over or make a new package for this driver?

I've not got the hardware so hard to test, or leave it for some poor soul to fix later?

**Comment by Mic92** [comment]
> linuxPackages_5_15.rtl8812au
> 
> is source from https://github.com/gordboy/rtl8812au-5.9.3.2

Are both projects still maintained? Does the new project only build with new kernels?

> 
> however that doesn't compile of 5.15 , however
> 
> https://github.com/morrownr/8812au-20210629
> 
> apparently does, however , do we swap over or make a new package for this driver?



> 
> I've not got the hardware so hard to test, or leave it for some poor soul to fix later?

If the original kernel maintainer has tested the code and it compiles we might not need to test it ourself.

**Comment by nixinator** [comment]
having looked at a few of these out of tree wireless drivers, the are quite a few repo that seem to support the same things, so it's hard to get a source of truth. Many have patches to get them into the more 'exotic modes' of operation, like this driver.


Although we have fairly good integration testing via Nixos tests , we need a test farm of real machines, connected  to real hardware, to do bare metal nixos testing in realtime!...  :-). I hear of projects in the embedded space, that do exactly that.

You can but dream(2nix) of something like this one day.

**Comment by Mic92** [comment]
If you don't have any real users to rely on, than we can also disable those drivers for newer kernel versions and wait until someone can try one of the forks.

**Comment by tomberek** [comment]
> Although we have fairly good integration testing via Nixos tests , we need a test farm of real machines, connected to real hardware, to do bare metal nixos testing in realtime!... :-). I hear of projects in the embedded space, that do exactly that.

Random comment/response: I did this a few years ago with some ARM boards and software-defined radios, using Hydra for orchestration. It was glorious!

**Comment by nixinator** [comment]
the force is strong with you @tomberek  , this needs to happen again, and become glorious once more! :-) 

'Nixos Certified Hardware'  :1st_place_medal: :1st_place_medal:

**Comment by MatthewCroughan** [comment]
@nixinator @tomberek 

Thankfully, I've been testing my Intel wireless card by using the NixOS Hardware Certification test suite. Looks like it passed on 5.14.15.

nixos-hardware-certification

**Pull request merged**